synapse/notifier.py

```diff
@@ -111,7 +112,9 @@ def __init__(
         with PreserveLoggingContext():
             self.notify_deferred = ObservableDeferred(defer.Deferred())
 
-    def notify(self, stream_key: str, stream_id: int, time_now_ms: int):
+    def notify(
+        self, stream_key: str, stream_id: Union[int, RoomStreamToken], time_now_ms: int,
```

Comment:
Ugh, so, argh. I spent some time trying to figure out what to do here, but basically we have a problem where the type of `stream_id` depends on the stream_key value. Adding overrides with literals sort works for the top level, but those functions then can't call other functions with literal overrides.

I think it may be possible to convert these into full `StreamTokens` with a bit of rejigging though

Reply:
I see. If this gets more complicated it might make sense to make a type that represents this (`StreamToken = Union[int, RoomStreamToken, FutureStreamToken]`).

Reply:
FWIW the new stuff will just be buried in `RoomStreamToken` so I'm hoping we don't see any more types for now. The big thing will be if we add types to the other streams

Reply:
> The big thing will be if we add types to the other streams

This is what I meant, yep!